### DISCUSSION TITLE
Welcome to MMM-Spotify Discussions!

### DISCUSSION THREAD

**Discussion by skuethe** [skuethe/MMM-Spotify#124] [General]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by Sheikh-A** [comment]
Hi All,

New to MM and trying to set one up, 
Could you explain where I put 
`
this.sendNotification("SPOTIFY_NEXT")
this.sendNotification("SPOTIFY_PAUSE")
`
Is this in the module file or config.js file?

**Reply by skuethe** [reply]
Hey there,

if you are using the build-in buttons to control MMM-Spotify, you do not need to use these notifications.
They give you the possibility to control MMM-Spotify from other modules. So for example, if you want to integrate with MMM-touch to perform a special gesture. This makes espacially sense for volume control, since there are no build-in buttons to in- / decrease volume. This is only possible via notifications.

**Comment by homeassistantettingen** [comment]
Hi i am new here and i am new in programming language...
I have the problem, that the Logo after the start pops up and then disappears. And nothing is shown.
Do you have any ideas what i am doing wrong?

Thanks Sascha

**Reply by skuethe** [reply]
Hey there, and welcome.

Please have a look at the following statement. There seems to be some misunderstanding what this module does and what not:
https://github.com/skuethe/MMM-Spotify/issues/125#issuecomment-758276331

You have to have some kind of control from external for this module to work. So either start playing music from some other device (mobile f.e.), or work with notification input to control it.
This module will not show on startup and give you full spotify control, like f.e. browse your playlists, select a song from it and start playing it.

I hope that clears things up. Just start playing a song from your mobile / pc / whatever and have a look at the module. It should now pick it up and give you control over the basic functions.

**Reply by homeassistantettingen** [reply]
Hi Thanks for the reply. 
I got it done, deleted everything and reinstaled. Now it works fine. Thanks
for the developing such a great addon!

Sascha

**Reply by skuethe** [reply]
great to hear that. Enjoy and if you have suggestions or other questions, don't hesitate to open issues or ask them here :)

**Comment by Opierarc** [comment]
Hello. Could you please explain the triggers behind notificationsOnSuspend and notificationsOnResume.  It doesn't matter what I do to spotify these notifications will not trigger, and by looking in the notification viewer it appears that they are not being sent by this module.

**Reply by skuethe** [reply]
Hi there.
To quote the example in the README:

```js
    // if you want to send custom notifications when suspending the module, f.e. switch MMM-Touch to a different "mode"
    notificationsOnSuspend: [
      {
        notification: "TOUCH_SET_MODE",
        payload: "myNormalMode",
      },
      {
        notification: "WHATEVERYOUWANT",
        payload: "sendMe",
      }
    ],
    // if you want to send custom notifications when resuming the module, f.e. switch MMM-Touch to a different "mode"
    notificationsOnResume: [
      {
        notification: "TOUCH_SET_MODE",
        payload: "mySpotifyControlMode",
      },
    ],
```

In these config settings you can define custom notifications which you want to send when this module (`MMM-Spotify`) is being suspended or resumed. Suspending and resuming are functions given by `MM²`. If you take a look [here](https://github.com/skuethe/MMM-Spotify/blob/6ab183a02e413c0f4ddc4ffb3a3e013d686c1422/MMM-Spotify.js#L1052-L1070), you can see that we overwrite these to use our `this.config.notificationsOnResume` and `this.config.notificationsOnSuspend` settings.

For more information about MM²s build in functions have a look at [the documentation](https://docs.magicmirror.builders/development/core-module-file.html#suspend).

So in the end, giving our example above, if the `MMM-Spotify` module gets suspended ("hidden"), we send two notifications:

1. `TOUCH_SET_MODE` with payload set to `myNormalMode`
2. `WHATEVERYOUWANT` with payload set to `sendMe`

and if we resume the `MMM-Spotify` module ("show"), then we send the following notification:

1. `TOUCH_SET_MODE` with payload set to `mySpotifyControlMode`


Now, it is up to you to then react on these notifications from within other modules. The idea behind these two settings is to give you control to communicate with other modules. Like the example says f.e. with `MMM-Touch`.

I hope that clarifies things.

**Comment by krillzon** [comment]
Hello, i tried running the node first_auth.js in order to create the token file, but i get this:
I made sure that i have the latest nodejs installed 

configurators.forEach((configurator) => {
                ^

TypeError: configurators.forEach is not a function

Great module, Thanks!

**Reply by skuethe** [reply]
Did you go through the installation instructions?
https://github.com/skuethe/MMM-Spotify#1-module-install
